### ISSUE TITLE
slicing in patterns

### ISSUE THREAD

**Issue by lutzhamel** [asteroid-lang/asteroid#39]
Asteroid should support slicing in patterns like so,
```
load "io".

let a = [3,5,7,4].
let a@[2,3] = a@[3,2].
println a. 

assert (a is [3,5,4,7]).
```

**Comment by olwmc** [comment]
Have we decided on a standard behavior for this? Should the length of the slice and the value you're trying to assign it as be forced to be the same?

I.e. would something like:
```
let a@[2,3] = a@[0]
```
be legal?

**Comment by lutzhamel** [comment]
The problem with this is that single element slices are easy because they kind of behave like patterns.  But what about this
```
let a = [1 to 10].
let a@[5 to 3] = a@[3 to 5].
```
All of a sudden we have active computation (computing the slice [5 to 3]) as part of a pattern.  I don't know what the solution is to that in the context of pattern matching.

To make it even weirder,
```
let a = [[1,2,3],[4,5,6],[7,8, 9]].
```
slicing only works row wise, but not column wise ... same restriction as on the right side.  

I would stay away from this for now.  When we have time we have to have a design session around perhaps later on this spring or early summer.

**Comment by olwmc** [comment]
Yeah exactly what I was thinking. Lots of weird edge cases! I'm happy to push this issue off for now.

**Comment by lutzhamel** [comment]
@olwmc ...now you got me thinking about this.  Having computations in the pattern is actually not unusual. Consider,
```
let x if x > 0 = some_val.
```
Given this,
```
let a@[5 to 3] = a@[3 to 5].
```
does not seem so outlandish.  It does get very interesting,
```
let b = [1 to 10].
let b@[1 to 5 step 2] = [1,2,3,4,5].
```

**Comment by lutzhamel** [comment]
@olwmc I did not answer your question from above.  The following,
```
let a@[2,3] = a@[0]
```
would not be legal because ultimately we are still considering this a pattern match.  That is, the structure on the right is matched against a pattern on the left. And here, the right is a single value and the pattern on the left is a pattern matching two values.  This match will fail...

**Comment by lutzhamel** [comment]
Here are some additional insights to array slicing on the left.  Array slicing on the left is not much different from any other structural pattern matching except that instead of variables being bound we have array locations being bound.  E.g.
```
let a = [1,2].
let [x,y] = a.
```
is not so different from,
```
let a = [1,2].
let a@[0,1] = a.
```
except that instead of variables being instantiated on the left we have array locations being instantiated.

Actually, we can view slicing as a short hand notation for pattern matching with lists and array elements as variables,
```
let a = [1,2].
let a@[0,1] = a.
```
is equivalent to
```
let a = [1,2].
let [a@0, a@1] = a.
```
With more general slices, assume
```
let a = [1 to 5].
let b = [1 to 10].
```
then
```
let b@[1 to 9 step 2] = a.
```
is equivalent to,
```
let b@[1,3,5,7,9] = a.
```
is equivalent to,
```
let [b@1,b@3,b@5,b@7,b@9] = a.
```
This might actually be an implementation strategy for slices because pattern matching on individual elements of an array already works.  So the last line actually works in Asteroid already!

@olwmc I take my earlier comments back, given the equivalence above I think implementing pattern matching on slices is not as daunting as I thought it would be if you want to give it a go!

**Comment by olwmc** [comment]
Resolved with #121
